### ISSUE TITLE
render tag

### ISSUE THREAD

**Issue by edgurgel** [edgurgel/solid#38]
https://shopify.dev/docs/themes/liquid/reference/tags/theme-tags#render

```
{% render 'snippet-name' %}
```

New tag!

**Comment by sb8244** [comment]
What are you thinking for what provides the template? Should there be a template provider option that can be passed in? I'm not sure that it should be part of the map data.

**Comment by edgurgel** [comment]
@sb8244 What about something like a module that implements a behaviour to load the snippets/files?

Then it would be passed as option just like it was done for custom tags?

https://github.com/edgurgel/solid/blob/8ffa4cc75d4ec9977850873a8cfee585e502fe54/test/integration/custom_tag_test.exs#L30-L32

`def render(template_or_text, values, options \\ [])`

**Comment by sb8244** [comment]
That makes a ton of sense regarding a module. I'm playing around with this a bit (finally) now that I need to implement a full templating system.

I'm coming at it from the angle of a custom tag, for now, to see how things work. Here's an example implementation that works on the surface but there's some really big issues with it:

```elixir
# DO NOT USE THIS CODE, IT IS BROKEN
  def home(conn, _params) do
    opts = [
      parser: Clove.Hubs.CustomParser,
      tags: %{
        "template" => Clove.Hubs.CustomParser.Template,
        "render" => Clove.Hubs.CustomParser.Template
      }
    ]

    vars = %{
      __templates: %{
        "home" => ~S({% render "test" %}),
        "test" => ~S(<strong>Test</strong>),
        "__missing_template" => "(render missing template)"
      },
      __opts: opts
    }

    html =
      ~S({% template "home" %} {% template %})
      |> Solid.parse!(opts)
      |> Solid.render(vars, opts)
      |> to_string()

    html(conn, html)
  end




defmodule Clove.Hubs.CustomParser.Template do
  def render(context, opts) do
    arguments = Keyword.get(opts, :arguments, [])
    name = Keyword.get(arguments, :value, "__missing_template")

    opts = context |> Map.from_struct() |> get_in([:vars, :__opts])
    template = context |> Map.from_struct() |> get_in([:vars, :__templates, name])

    %Solid.Template{parsed_template: text} = Solid.parse!(template, opts)

    Solid.render(text, context, opts)
    |> elem(0)
    |> to_string()
  end
end
```

This is my attempt at allowing recursive templates to be used. However, it has a glaring issue if I say `{% render "home" %}` as the home template...infinite loop.

I'm working through different options for that now. Let me know if this approach is just way off.

**Comment by edgurgel** [comment]
Hmm interesting. Liquid avoids infinite loops by relying on max depth of tags I think? https://github.com/Shopify/liquid/blob/e6eef4b2c40f54d03aceffd0d8ab5f2da90db5b7/lib/liquid/block.rb#L73

Not sure if/how we would do this but that's an interesting way to avoid this?

Maybe don't worry about infinite loops for now and we tackle this max depth as a separate issue?

**Comment by sb8244** [comment]
I haven't looked at making this generic yet. But, it's working great for me so far! No snags. Gathering some thoughts:

* My implementation right now requires passing all of the templates into the context. I think that a provider function pattern could be used, where Solid will invoke a function with the context it has. The implementor could decide to fetch from that context, or could fetch from a database / cache
* Agreed on not worrying about infinite loops. If the custom tag can modify the context (brought up in another issue), then we could keep track of max depth and detect infinite loops. I imagine it would look much like the Ruby code, a `depth` context attribute that is incremented as it goes down

**Comment by edgurgel** [comment]
Closed by #85
